### ISSUE TITLE
Reference parameter in virtual method doesn't update correctly when class is mocked

### ISSUE THREAD

**Issue by iitaiy** [moq/moq4#585]
I have test that check ref parameter value from event that raised after the change. The test work fine when I create real class but when I try mocked class its fail. Simple code example:

    [TestClass]
    public class Fixture
    {
        [TestMethod]
        public void CheckRealClass()
        {
            // this test passed
            var classToTest = new NumberChangerClass();
            CheckClass(classToTest);
        }

        [TestMethod]
        public void CheckMockedClass()
        {
            // this test failed
            var classToTest = new Mock<NumberChangerClass>() { CallBase = true };
            CheckClass(classToTest.Object);
        }

        private void CheckClass(NumberChangerClass numberChangerClass)
        {
            // Arrange
            var number = 1;
            numberChangerClass.StorageChanged += (sender, args) =>
            {
                Assert.AreEqual(2, number);
            };

            // Act
            numberChangerClass.SetValue(ref number, 2);
        }
    }

    public class NumberChangerClass
    {
        public event EventHandler StorageChanged;

        public virtual void SetValue(ref int storage, int newValue)
        {
            storage = newValue;
            StorageChanged?.Invoke(this, null);
        }
    }

In `CheckMockedClass` the `number` is actually changed but only after the the code exits from `SetValue` method.

Thank you for the help.

**Comment by stakx** [comment]
Near-duplicate of #466.

@iitaiy, the behaviour you are observing is a direct consequence of how DynamicProxy (which Moq depends on) handles by-reference parameters during method call interception (which is documented at https://github.com/castleproject/Core/blob/master/docs/dynamicproxy-by-ref-parameters.md): Arguments are buffered in a separate location during call interception and only written back at the end of interception, that's why `ref`parameters are updated in a delayed / deferred manner. There is unfortunately nothing that can be done about it.

Thank you for taking the time to report this, though!
